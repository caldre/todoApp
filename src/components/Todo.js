import React, {useState} from "react";

const Todo = (props) => {
  const [hovered, setHovered] = useState(false);

  const { todoDetails, toggleCompleteStatus, removeTodo, darkMode } = props;

const renderCompleteIcon = () => {
  if (todoDetails.complete){
  return  <i id="check" className={`fas fa-check-circle fa-2x`}/>
}
  else return <i id="check"/>
}

const setComplete = () => {
  if(todoDetails.complete) {
    return "completed"
  } else return "incomplete"
}

  const handleTogglingClick = () => {
    toggleCompleteStatus(todoDetails.id);
  };

  const handleRemoveClick = () => {
    removeTodo(todoDetails.id);
  };

  const setTheme = () => {
    if (darkMode) {
      return "dark-mode"
    } else return "light-mode"
  }

  return (
    <div    className={`todos-wrapper ${setTheme()}`}
            onMouseOver={() => setHovered(true)}
            onMouseLeave={() => setHovered(false)}>
              {renderCompleteIcon()}
      <div  className={`todo-text ${setComplete()}`}
            onClick={handleTogglingClick}  >{todoDetails.name}</div>
      
      {hovered ? <i id="delete-btn" className={`fas fa-window-close fa-2x`} onClick={handleRemoveClick}></i> : <i id="delete-btn"/>}
      
    </div>
  );
};

export default Todo;
