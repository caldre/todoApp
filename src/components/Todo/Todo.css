.todos-wrapper {
  display: flex;
  justify-content: space-between;
  min-height: 3rem;
  margin-top: 1rem;
  cursor: pointer;
  box-shadow: var(--box-shadow) 2px 2px 10px;
}

.todos-wrapper.light-mode {
  background-color: var(--light-bg-color);
  transition: var(--fade-time);
}

.todos-wrapper.light-mode:hover {
  background-color: var(--light-hovered);
  transition: var(--hover-time);
}

.todos-wrapper.dark-mode {
  background-color: var(--dark-bg-color);
  transition: var(--fade-time);
}

.todos-wrapper.dark-mode:hover {
  background-color: var(--dark-hovered);
  transition: var(--hover-time);
}

.btn.complete-btn {
  padding: 0rem 0.5rem;
  width: 100%;
  align-self: stretch;
}

.todo-text-wrapper {
  display: flex;
  align-items: center;
}

.todo-text {
  flex-grow: 4;
  text-align: center;
  padding: 1rem 0rem;
}

.todo-text.completed {
  transition: var(--fade-time);
  text-decoration: line-through;
  opacity: 0.3;
}

.btn.delete-btn {
  align-self: center;
  margin: 0rem 0.5rem;
  flex-grow: 1;
  min-width: 2rem;
}

.check-icon {
  transition: var(--icon-transition);
  padding: 0rem 0.5rem;
  min-width: 2rem;
  color: var(--success);
  -webkit-text-stroke-width: 0.5px;
  -webkit-text-stroke-color: rgba(0, 0, 0, 1);
}

.delete-icon {
  transition: var(--icon-transition);
  color: var(--danger);
  -webkit-text-stroke-color: rgba(0, 0, 0, 1);
}
