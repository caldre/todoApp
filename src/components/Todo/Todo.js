import React, { useState } from "react";
import AddTodo from "../AddTodo/AddTodo";
import "./Todo.css";

const Todo = (props) => {
  const [hovered, setHovered] = useState(false);
  const [editing, setEditing] = useState(false);

  const { toggleCompleteStatus, removeTodo, moveItem } = props;
  const { userId, id, title, completed } = props.todoDetails;

  const handleTogglingClick = () => {
    toggleCompleteStatus(id);
  };

  const handleRemoveClick = () => {
    removeTodo(id);
  };

  return editing ? (
    <AddTodo todoDetails={props.todoDetails} onSubmit={props.onSubmit} />
  ) : (
    <div
      className={`todo-wrapper`}
      onMouseOver={() => setHovered(true)}
      onMouseLeave={() => setHovered(false)}
      data-testid="todo-wrapper"
    >
      <button
        className="btn complete-btn"
        onClick={handleTogglingClick}
        aria-label="Toggle tasks complete status"
        data-testid="toggle"
      >
        <div className="todo-text-wrapper">
          <i
            data-testid="check-icon"
            className={`check-icon ${
              completed ? "fas fa-check-circle fa-2x" : ""
            }`}
          />

          <p className={`todo-text ${completed ? "completed" : ""}`}>{title}</p>
        </div>
      </button>
      <div className="todo-tools-wrapper">
        <div className="arrow-wrapper">
          <button className="btn btn-up" onClick={() => moveItem(id, -1)}>
            <i className={`down-icon ${hovered ? "fas fa-caret-up" : ""}`}></i>
          </button>
          <button className="btn btn-down" onClick={() => moveItem(id, +1)}>
            <i
              className={`down-icon ${hovered ? "fas fa-caret-down" : ""}`}
            ></i>
          </button>
        </div>
        <button className="btn btn-edit" onClick={() => setEditing(true)}>
          <i className={`edit-icon ${hovered ? "fas fa-edit fa-2x" : ""}`}></i>
        </button>

        <button
          className="btn delete-btn"
          onClick={handleRemoveClick}
          aria-label="Delete task"
          data-testid="delete"
        >
          <i
            data-testid="delete-icon"
            className={`delete-icon ${
              hovered ? "fas fa-window-close fa-2x" : ""
            }`}
          ></i>
        </button>
      </div>
    </div>
  );
};

export default Todo;
