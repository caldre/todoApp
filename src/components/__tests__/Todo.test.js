import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Todo from "../Todo/Todo";
import renderer from "react-test-renderer";

const expectedId = 12341234;
let props = { id: expectedId, name: "Go to the supermarket", complete: false };

test("Todo component renders", () => {
  const component = render(<Todo todoDetails={props} />);

  expect(component.container.firstChild).toMatchSnapshot();
});

test("Check that component renders with correct default props", () => {
  const testRenderer = renderer.create(<Todo todoDetails={props} />);
  const testInstance = testRenderer.root;

  expect(testInstance.findByType(Todo).props.todoDetails.id).toBe(expectedId);
  expect(testInstance.findByType(Todo).props.todoDetails.name).toBe(
    "Go to the supermarket"
  );
  expect(testInstance.findByType(Todo).props.todoDetails.complete).toBe(false);
});

test("Toggles delete-icon with mouse hover", () => {
  const component = render(<Todo key={props.id} todoDetails={props} />);
  const todoWrapper = component.getByTestId("todo-wrapper");
  const deleteIcon = component.getByTestId("delete-icon");

  expect(deleteIcon.classList.contains("fa-window-close")).toBe(false);
  fireEvent.mouseOver(todoWrapper);
  expect(deleteIcon.classList.contains("fa-window-close")).toBe(true);
  fireEvent.mouseLeave(todoWrapper);
  expect(deleteIcon.classList.contains("fa-window-close")).toBe(false);
});

test("Toggle button is called exactly once with a correct argument", () => {
  const mockToggle = jest.fn();
  const component = render(
    <Todo todoDetails={props} toggleCompleteStatus={mockToggle} />
  );
  const toggleComplete = component.getByTestId("toggle");

  expect(mockToggle).toHaveBeenCalledTimes(0);
  fireEvent.click(toggleComplete);
  expect(mockToggle).toHaveBeenCalledTimes(1);
  expect(mockToggle).toHaveBeenCalledWith(expectedId);
});

test("Delete button is called exactly once with a correct argument", () => {
  const mockDelete = jest.fn();
  const component = render(
    <Todo key={props.id} todoDetails={props} removeTodo={mockDelete} />
  );
  const deleteButton = component.getByTestId("delete");

  expect(mockDelete).toHaveBeenCalledTimes(0);
  fireEvent.click(deleteButton);
  expect(mockDelete).toHaveBeenCalledTimes(1);
  expect(mockDelete).toHaveBeenCalledWith(expectedId);
});
