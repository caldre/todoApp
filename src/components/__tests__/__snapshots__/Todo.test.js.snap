// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Changes state with mouse events 1`] = `
<div
  className="todos-wrapper undefined"
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <button
    className="btn complete-btn"
    onClick={[Function]}
  >
    <div
      className="todo-text-wrapper"
    >
      <i
        className="check-icon"
      />
      <p
        className="todo-text incomplete"
      >
        Go to the supermarket
      </p>
    </div>
  </button>
  <button
    className="btn delete-btn"
    onClick={[Function]}
  >
    <i
      className="delete-icon"
    />
  </button>
</div>
`;

exports[`Changes state with mouse events 2`] = `
<div
  className="todos-wrapper undefined"
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <button
    className="btn complete-btn"
    onClick={[Function]}
  >
    <div
      className="todo-text-wrapper"
    >
      <i
        className="check-icon"
      />
      <p
        className="todo-text incomplete"
      >
        Go to the supermarket
      </p>
    </div>
  </button>
  <button
    className="btn delete-btn"
    onClick={[Function]}
  >
    <i
      className="delete-icon"
    />
  </button>
</div>
`;

exports[`Changes state with mouse events 3`] = `
<div
  className="todos-wrapper undefined"
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <button
    className="btn complete-btn"
    onClick={[Function]}
  >
    <div
      className="todo-text-wrapper"
    >
      <i
        className="check-icon"
      />
      <p
        className="todo-text incomplete"
      >
        Go to the supermarket
      </p>
    </div>
  </button>
  <button
    className="btn delete-btn"
    onClick={[Function]}
  >
    <i
      className="delete-icon fas fa-window-close fa-2x"
    />
  </button>
</div>
`;

exports[`Todo component renders 1`] = `
<div
  className="todos-wrapper undefined"
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <button
    className="btn complete-btn"
    onClick={[Function]}
  >
    <div
      className="todo-text-wrapper"
    >
      <i
        className="check-icon"
      />
      <p
        className="todo-text incomplete"
      >
        Go to the supermarket
      </p>
    </div>
  </button>
  <button
    className="btn delete-btn"
    onClick={[Function]}
  >
    <i
      className="delete-icon"
    />
  </button>
</div>
`;
